File: mobile/src/services/SearchService.js
import Shop from '../models/Shop';
import Product from '../models/Product';
import ApiService from './ApiService';

class SearchService {
  async search(keyword) {
    const data = await ApiService.get(
      `/search?keyword=${encodeURIComponent(keyword)}`,
    );
    return {
      shops: (data.shops || []).map((item) => new Shop(item)),
      products: (data.products || []).map((item) => new Product(item)),
    };
  }
}

export default new SearchService();

File: server/src/main/java/com/foodsaver/service/SearchService.java
package com.foodsaver.service;

import com.foodsaver.dto.ProductResponse;
import com.foodsaver.dto.SearchResponse;
import com.foodsaver.dto.ShopResponse;
import com.foodsaver.repository.ProductRepository;
import com.foodsaver.repository.ShopRepository;
import java.util.List;
import java.util.stream.Collectors;
import org.springframework.stereotype.Service;

@Service
public class SearchService {

    private final ShopRepository shopRepository;
    private final ProductRepository productRepository;

    public SearchService(ShopRepository shopRepository,
            ProductRepository productRepository) {
        this.shopRepository = shopRepository;
        this.productRepository = productRepository;
    }

    public SearchResponse search(String keyword) {
        List<ShopResponse> shops = shopRepository.searchByKeyword(keyword).stream()
                .map(ShopResponse::new)
                .collect(Collectors.toList());

        List<ProductResponse> products = productRepository
                .searchByKeyword(keyword).stream()
                .map(ProductResponse::new)
                .collect(Collectors.toList());

        return new SearchResponse(shops, products);
    }
}
